advanced-data-types.md

```diff
@@ -0,0 +1,195 @@
+
+## Constructor validation
+
+One of the core requirements of Realm is that all models need to have a parameterless constructor. This is needed because Realm needs to be able to instantiate an object without figuring out what the arguments to pass to the constructor. What not many people know is that you can make this parameterless constructor private to communicate expectations  to your callers. This means that if you have an `Person` class where you absolutely expect that a `Name` is provided upon object creation, you can have a public constructor with a `name` argument and a private parameterless one for use by Realm:
+
+```csharp
+class Person : RealmObject
+{
+    public string Name { get; set; }
+
+    public Person(string name)
+    {
+        ValidateName(name);
+
+        Name = name;
+    }
+
+    // This is used by Realm, even though it's private
+    private Person()
+    {
+    }
+}
+```
+
+And I know what some of you may be thinking - "oh no 😱 does that mean Realm uses the suuuuper slow reflection to create object instances". Fortunately, the answer is "no". Instead, at compile-time, Realm injects a nested helper class in each model that has a `CreateInstance` method. Since the helper class is nested in the model classes, it has access to private members and is thus able to invoke the private constructor.
+
+## Property access modifiers
+
+Similar to the point above - another relatively unknown feature of Realm is that persisted properties don't need to be public. You can either have the entire property be private or just one of the accessors. This synergizes nicely with the private constructor technique that we mentioned above. If you expose a constructor that explicitly validates the person's name, it would be fairly annoying to do all that work and have some code accidentally set the property to `null` the very next line. So it would make sense to make the setter of the name property above private:
+
+```csharp
+class Person : RealmObject
+{
+    public string Name { get; private set; }
+
+    public Person(string name)
+    {
+        // ...
+    }
+}
+```
+
+That way you're communicating clearly to the class consumers that they need to provide the name at object creation time and that it can't be changed later. A very common use case here is to make the `Id` setter private and generate a random `Id` at object creation time:
+
+```csharp
+class Transaction : RealmObject
+{
+    public Guid Id { get; private set; } = Guid.NewGuid();
+}
+```
+
+Sometimes it makes sense to make the entire property private - typically when you want to expose a different public property that wraps it. If we go back to our `Person` and `Name` example - perhaps we want to allow changing the name, but we want to still validate the new name before we persist it. Then, we create a private autoimplemented property that Realm will use for persistence, and a public one that does the validation:
```

Comment:
Do we actually offer a way to combine them yet?

Reply:
and by that I mean: do we have plans to do that?
Would also solve LINQ and notification problem I suppose.

Reply:
What do you mean?

Reply:
Gotcha - it's not possible for now, but we do hope to support it.